Add local Schur indices for a character (Unger's algorithm)

ThomasBreuer opened the pull request.

ThomasBreuer committed:
local Schur indices for a character (Unger's alg.)

thofma commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,527 @@
+  # such that q divides p-1 and G contains an element of order p q.
+  for pair in factor(u)
+    l = pair[1]
+    e = pair[2]
```
```suggestion
  for (l, e) in factor(u)
```

lgoettgens commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,527 @@
+
+add_verbosity_scope(:SchurIndices)
```
This should be moved to the `__init__` function in `src/Oscar.jl`

lgoettgens commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,527 @@
+
+This example is described in the paper
+  T. Yamada, Schur indices over the 2-adic field,
+  Pacific Journal of Mathematics 74, no. 1, 1978.
```
better to put this into the bibliography and cite here

lgoettgens commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,527 @@
+  is_abelian(X) && return no_type
+  (order(X) == 2*order(Z)) || return no_type
+  (mod(twopart, div(order(P), order(X))) == 0) || return no_type
+  return("QD", r)
```
```suggestion
  return ("QD", r)
```

lgoettgens commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,527 @@
+  end
+  @vprintln :SchurIndices 1 "step 4, return $result"
+  return result
+end;
```
```suggestion
end
```

ThomasBreuer committed:
address comments

fieker commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,520 @@
+  @req is_odd(p) && is_prime(p) "p must be an odd prime integer"
+  @req mod(p-1, d) == 0 "d must divide p-1"
+
+  r = GAP.Globals.PrimitiveRootMod(p)
```
We have that functionality on Hecke, somewhere. I think (independent of this PR) we need to make it available

ThomasBreuer replied:
Yes, there is `Hecke.gen_mod_pk`.

fieker commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,520 @@
+  if length(vals) == 0
+    F = GAP.Globals.Rationals
+  else
+    F = GAP.Globals.Field(GAP.Globals.Rationals, vals)
```
we need to talk. givem the amout of GapField stuff in here, we seem to be missing canonical useful stuff in Oscar direct....I really glad it works, but here there seems to be work required (pos. by me)

fieker commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,520 @@
+
+julia> local_schur_indices(chi)
+2-element Vector{Pair{Int64, Int64}}:
+ -1 => 2
```
this too needs thought: for G-Modules, the local indices can be labled by prime numbers as all conjugate prime ideals behave the same (a property of the Schur-Subgroupof the Brauer group). In general there need to be prime ideals or places here if we want to extend/ use the same interface for A-modules

thofma replied:
I think the unqualified `local_schur_indices(chi)` for complex characters need not be in sync with the modules over fin.-dim. algebras. The interface here is the same as in Magma, except that Magma uses 0 instead of -1 for the archimedean place. I am leaning towards `0`, since we use this convention already for things like `hilbert_symbol`.

ThomasBreuer replied:
Concerning the 0 vs. -1:
I have tried to use the same format as in `experimental/GModule/src/GModule.jl`. I have no preference, it should just be consistent.

thofma replied:
@fieker are you opposed to changing 0 to -1?

fingolfin commented on docs/oscar_references.bib:
```diff
@@ -2732,6 +2732,19 @@ @Article{Yam18
+  pages         = {273--275},
+  year          = {1978},
+  url           = {http://projecteuclid.org/euclid.pjm/1102810458}
+}
```
This PR talks about "Unger's algorithm", which makes me wonder why there is no .bib entry for a paper by Unger added here?

fingolfin commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,520 @@
+  return P, tbl[i]
+end 
+
+# See section 5 in the paper by Unger:
```
If that papers was in the bib we could write something like "See section 5 of `[UngXY]`" here

fingolfin commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,520 @@
+  facts = sort(collect(factor(n)))
+  length(facts) == 2 || return no_type
+  (facts[1][1] == 2 && facts[2][2] == 1) || return no_type
+  r = facts[2][1]
```
```suggestion
  is_even(n) || return no_type
  _, r = remove(n, 2)
  is_prime(r) || return no_type
```

fingolfin commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,520 @@
+    # Compute the conductor of the largest cyclotomic field
+    # that is contained in the character field of `chi`.
+    gapfield = GAPWrap.Field(GapObj(chi))
+    N = GAPWrap.Conductor(gapfield)
```
I was confused by this because I read "Compute the conductor" as referring to the value `N`... but I think it's really talking about `cond`. How about this to make it clearer?
```suggestion
    # Compute the conductor `N` of the smallest cyclotomic field
    # that contains the character field of `chi`.
    gapfield = GAPWrap.Field(GapObj(chi))
    N = GAPWrap.Conductor(gapfield)
    # Compute the conductor `cond` of the largest cyclotomic field
    # that is contained in the character field of `chi`.
```

fingolfin commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,520 @@
+      GAPfield_p[p] = _Q_p_chi(GapObj(chi), p)
+      # Run over the p-regular elements (without the identity element).
+      pregs = findall(x -> mod(x, p) != 0 && x != 1, ords)
+      vals = filter(x -> isa(x, GapObj), GAP.Obj[x for x in GapObj(chi)])
```
what does this `filter` statement do, what values in there are not of type `GapObj`? Maybe some are `Int` and thus can be ignored? Would be good to state this, it seems somewhat magical otherwise.

ThomasBreuer replied:
I have added a comment line. (GitHub does not show the "Outdated" flag.)

fingolfin commented on src/Groups/schur_index.jl:
```diff
@@ -0,0 +1,520 @@
+#T `modord(N, 1)` could return 1 (smallest pos. int. i s.t. 1 divides (N^i-1))
+    stab = Int[]
+    sigma_p = mod(1 + a*(p-1)*ppart, N)
+    res = ppart == 1 ? [1] : Vector{Int}(GAP.Globals.PrimeResidues(GapObj(ppart)))
```
@fieker @thofma do we have a function like `PrimeResidues` in OSCAR?

> `PrimeResidues(m)`  returns the set of integers from the range `[ 0 .. Abs( m )-1 ]` that are coprime to the integer `m`.

fingolfin reviewed (commented):
Thank you, great to see this!

I left some minor comments (hopefully not too stupid).

ThomasBreuer committed:
add `central_character`, `defect_group`, `power_map`

ThomasBreuer committed:
next iteration

- address comments
- decide about cyclic defect groups of blocks without using the group

ThomasBreuer committed:
write `Oscar.prime_residues` in doctests

ThomasBreuer committed:
fix a test

thofma commented:
The function `prime_residues` looks useful and should be added to Hecke. Would it be ok to rename it to  `coprime_residues` in this case?

ThomasBreuer commented:
> The function `prime_residues` looks useful and should be added to Hecke. Would it be ok to rename it to `coprime_residues` in this case?

No problem with this name. Even `Hecke.gen_mod_pk` is an acceptable name.

thofma commented:
> > The function `prime_residues` looks useful and should be added to Hecke. Would it be ok to rename it to `coprime_residues` in this case?
> 
> No problem with this name. Even `Hecke.gen_mod_pk` is an acceptable name.

no, it is an abomination

edit: I opened https://github.com/thofma/Hecke.jl/issues/1786 to keep track of this

fieker commented:
On Fri, Feb 28, 2025 at 06:22:22AM -0800, Tommy Hofmann wrote:
> @thofma commented on this pull request.
> 
> 
> 
> > +Return the array of pairs `p => m_p(chi)` such that `m_p(chi) > 1` holds.
> +The Schur index over the real field is stored at `p = -1` if applicable.
> +
> +`cyclic_defect` contains primes `p` such that `chi` is known to belong to
> +a `p`-block of cyclic defect.
> +
> +# Examples
> +```jldoctest
> +julia> G, chi = Oscar.yamada_example(5, 2);
> +
> +julia> order(G)
> +40
> +
> +julia> local_schur_indices(chi)
> +2-element Vector{Pair{Int64, Int64}}:
> + -1 => 2
> 
> @fieker are you opposed to changing 0 to -1?
> 
No, fine with me. I was more thinking about a possibility of also doing
s.th. with places. but for now, I prefer -1 over 0
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/4628#discussion_r1975496256
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

thofma commented:
Sorry, I meant `0` *instead* of `-1`. We already use it for other things like
```
julia> hilbert_symbol(2, 3, 0)
1
```
But Claus is also happy with using `0`, as he confirmed a few minutes ago in the meeting.

ThomasBreuer commented:
O.k., then I change the format for the output of `local_schur_indices` in the current pull request and in `experimental/GModule/src/GModule.jl`.

ThomasBreuer committed:
change output format of `local_schur_indices`

ThomasBreuer committed:
fix more `-1` -> `0` issues

ThomasBreuer committed:
unify `schur_index` and `local_schur_indices`

Let `schur_index` for characters delegate to `local_schur_indices`
instead of calling a function from the GAP package wedderga,
except in the case of indicator -1, where we conclude Schur index 2
without computing the local indices.

ThomasBreuer committed:
fix reference to docstring
